src/Type/User/Mutation/UserMutation.php

```diff
@@ -240,35 +243,66 @@ public static function update_additional_user_object_data( $user_id, $input, $mu
 	 * @param array $roles   List of roles that need to get added to the user
 	 *
 	 * @access private
+	 * @throws /\Exception
 	 */
 	private static function add_user_roles( $user_id, $roles ) {
 
-		if ( empty( $roles ) || ! is_array( $roles ) ) {
+		if ( empty( $roles ) || ! is_array( $roles ) || ! current_user_can( 'edit_user', $user_id ) ) {
 			return;
 		}
 
 		$user = get_user_by( 'ID', $user_id );
 
 		if ( false !== $user ) {
+
 			foreach ( $roles as $role ) {
-				self::verify_user_role( $role );
-				$user->add_role( $role );
+
+				$verified = self::verify_user_role( $role, $user_id );
+
+				if ( true === $verified ) {
+					$user->add_role( $role );
+				} else if ( is_wp_error( $verified ) ) {
+					$message = $verified->get_error_message();
+					throw new \Exception( $message );
+				} else if ( false === $verified ) {
+					throw new \Exception( __( 'This role cannot be added to this user', 'wp-graphql' ) );
+				}
+
 			}
 		}
 
 	}
 
 	/**
-	 * Method to check if the user role is valid, and if the current user has permission to add, or remove it from a
-	 * user.
+	 * Method to check if the user role is valid, and if the current user has permission to add, or
+	 * remove it from a user.
 	 *
-	 * @param string $role Name of the role trying to get added to a user object
+	 * @param string $role    Name of the role trying to get added to a user object
+	 * @param int    $user_id The ID of the user being mutated
 	 *
-	 * @return bool
-	 * @throws \Exception
+	 * @return mixed|bool|\WP_Error
 	 * @access private
 	 */
-	private static function verify_user_role( $role ) {
+	private static function verify_user_role( $role, $user_id ) {
+
+		global $wp_roles;
+
+		if ( ! isset( $wp_roles->role_objects[ $role ] ) ) {
+			return new \WP_Error( 'wpgraphql_user_invalid_role', sprintf( __( 'The role %s does not exist.' ), $role ) );
+		}
+
+		$potential_role = $wp_roles->role_objects[ $role ];
```

Comment:
Could probably reduce some duplicate code here by making the above chunk:
```php
$potential_role = ( isset( $wp_roles->role_objects[ $role ] ) $wp_roles->role_objects[ $role ] : '';
if ( empty( $potential_role ) ) {
    return new \WP_Error( ... );
}
```
You don't have to change it, the above just feels a little cleaner to me.